vector/src/main/java/im/vector/app/features/home/NewHomeDetailFragment.kt

```diff
@@ -449,10 +437,11 @@ class NewHomeDetailFragment @Inject constructor(
         return this
     }
 
-    override fun onBackPressed(toolbarButton: Boolean) = if (spaceStateHandler.getCurrentSpace() != null) {
-        navigateBack()
+    override fun onBackPressed(toolbarButton: Boolean) = try {
+        val lastSpace = spaceStateHandler.popSpaceBackstack()
+        spaceStateHandler.setCurrentSpace(lastSpace, isForwardNavigation = false)
         true
-    } else {
+    } catch (e: NoSuchElementException) {
```

Comment:
which part of the api throws `NoSuchElementException`? if we're using exceptions for control flow we'll probably want to document it in the `SpaceStateHandler` api and add a unit test case

ideally we wouldn't rely on exceptions by catching and modelling the state within our own domain, unless this is an exceptional case out of our control :thinking:

Reply:
It happens in `SpaceStateHandlerImpl:144` (`val poppedSpaceId = (spaceBackstack).removeLast`)

The reason I wanted to use exception here instead of null is because the space backstack can include a valid `null` which signifies the "All Chats" space, thus we can't use null as a fail option

Reply:
my concern with using the exception is that it reads as if an error has occurred when in reality the user is at the top of the backstack (an empty stack), there's the risk that a `NoSuchElementException` could be thrown somewhere else and we unexpectedly catch it, causing swallowed errors :thinking: 

a named guard function provides extra context in relation to the navigation and allows the user of valid nulls 

```kotlin
fun onBackPressed() = when {
  spaceStateHandler.isRoot() -> false
  else -> {
        val lastSpace = spaceStateHandler.popSpaceBackstack()
        spaceStateHandler.setCurrentSpace(lastSpace, isForwardNavigation = false)
        true
  }
}

private fun SpaceStateHandler.isRoot() = getBackstack().isEmpty()

```

Reply:
Good suggestion. Changes made